### ISSUE TITLE
Distributional shift when migrating from PyTorch to Albumentations

### ISSUE THREAD

**Issue by ananiask8** [albumentations-team/albumentations#1022]
## 🐛 Bug

When migrating from PyTorch transforms to Albumentations, the predictions given by the model change.

## To Reproduce

I essentially followed the recommended migration policy given in the [docs](https://albumentations.ai/docs/examples/migrating_from_torchvision_to_albumentations/).

My code looks as follows:

## With PyTorch transforms

```python
from torch.utils.data import Dataset
from torchvision import transforms as t

class CustomDataset(Dataset):
    def __init__(self, opts):
        # ...
        self.transform = t.Compose([
	        t.Resize((opts.input_height, opts.input_width)),
	        t.ToTensor(),
	        t.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
        ])

    # ...

    def __getitem__(self, item: int):
        label = self.data.iloc[item].label
        name = self.data.iloc[item].img_name
        img_transformed = img = Image.open(join_path(self.dataset_path, name))
        if self.transform is not None:
            img_transformed = self.transform(img)
       return img_transformed, label
```

## With Albumentations

```python
from torch.utils.data import Dataset
from albumentations.pytorch import transforms as Ap
import albumentations as A

class CustomDataset(Dataset):
    def __init__(self, opts):
        # ...
        self.transform = A.Compose([
	        A.Resize(opts.input_height, opts.input_width),
	        A.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
                Ap.ToTensorV2()
        ])
   
    # ...

    def __getitem__(self, item: int):
        label = self.data.iloc[item].label
        name = self.data.iloc[item].img_name
        img_transformed = img = cv2.imread(join_path(self.dataset_path, name))
        img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
        if self.transform is not None:
            img_transformed = self.transform(image=img)['image']
       return img_transformed, label
```

## Expected behavior

Assuming this is due to some changes in the distribution introduced by the different transforms, it would be understandable if a guarantee cannot be provided against this issue. However, this would be more understandable when more complicated transformations are involved. In this case, since the transforms don't even include random values, it seems to me that there could be a mistake on my side. Since I followed the provided migration guide, if I made a mistake, it is likely this should be documented in the guide.

## Environment

 - PyTorch v1.9.0
 - Albumentations v1.0.3
 - Python v3.7.11
 - macOS

**Comment by Dipet** [comment]
What difference did you find?  What loss do you get with torchvision transforms and albumentations?
I see only one difference - OpenCV could produce different artifacts when resizing images. It is a known issue https://arxiv.org/pdf/2104.11222.pdf

**Comment by BloodAxe** [comment]
There are two things that may cause this change:

- Different resize implementation that @Dipet pointed out. This indeed could introduce a small difference in the resized images. In Albumentations we prefer OpenCV over PIL for speed.
- Different image decoding. In PyTorch example you're using PIL's Image.open, while we again use OpenCV for the sake of speed. Different implementations of jpeg decoders (libjpeg, libjpeg-turbo) can and will produce slightly different images. 

I honestly don't think it is a big issue. If we look at NVidia DALI as preprocessing/augmentation pipeline, they will produce images that will differs from both OpenCV & PIL results. On the other hand, applying different image decoding and resizing methods at train time can be seen as additional augmentation scheme, where you enforce model to be robust to those distribution shifts.

**Comment by ananiask8** [comment]
The difference was not dramatic: a drop of 4% on the test partition of a small dataset (test partition with around 250 data points), which is about 10 extra misclassifications. I would guess that this is happening in data which was _difficult_ for the model (i.e., top2 classes have similar probability). Feel free to close if this is something you already had under your radar; I thought to let you know in case this wasn't expected behavior.
